lib/src/git.rs

```diff
@@ -1324,66 +1331,166 @@ fn push_refs(
+                    .get(dst_refname)
+                    .expect("Push is trying to move a ref it wasn't asked to move"))
+                .clone();
+                let expected_location_oid = expected_location
+                    .map(|id| git2::Oid::from_bytes(id.as_bytes()).unwrap())
+                    .unwrap_or(git2::Oid::zero());
+
+                // `update.src()` is the current actual position of the branch
+                // `dst_refname` on the remote. `update.dst()` is the position
+                // we are trying to push the branch to (AKA our local branch
+                // location). 0000000000 is a valid value for either, indicating
+                // branch creation or deletion.
+                if update.src() != expected_location_oid {
+                    // There are a few corner cases where we allow the push even
+                    // though the branch is in an unexpected location. All of
+                    // these exceptions **require** that the push is a
+                    // "fast-forward" on the remote: the push destination is a
+                    // descendant of or equal to the unexpected location. There
+                    // are additional conditions on `expected_location` that
+                    // need to be satisfied; see the `if`` conditions below.
+                    //
+                    // This is sometimes more restrictive than Git's `git push`
+                    // which would allow *all* fast-forwards to succeed,
+                    // regardless of where the expected location is. The
+                    // difference seems like it would matter rarely in practice.
+                    // Our rule is also different from Git's `git push
+                    // --force-with-lease`, which would AFAIK not allow any
+                    // corner cases at all. TODO(ilyagr): Double-check the last
+                    // statement.
+                    //
+                    // The reasoning for our rule is that we can consider this
+                    // as a branch conflict with 2 sides, `update.dst()` (our
+                    // branch location) and `update.src()` (their branch
+                    // location), and with `expected_location` as the base of
+                    // the conflict. In fact, if the user performed a `jj git
+                    // fetch` instead of `jj git push` they are performing, they
+                    // would get exactly this branch conflict. We decide to
+                    // allow the push **if and only if** `jj` would auto-resolve
+                    // this conflict to our location, `update.dst()`.
+                    // Conveniently, this also means that the push will be
+                    // successful **if and only if** `jj git fetch` is a no-op
+                    // for this branch (and, in particular, would not create a
+                    // conflict).
```

Comment:
Out of curiosity, have you considered actually using our `merge_ref_targets()` logic? I think that's doable with some sanity check for unknown commit ids. Will it simplify the push negotiation?

Reply:
I thought about it, and then decided to keep this equivalence inside tests.

I think the behavior will be easier to debug and explain if it's written out. For example, this way we can have tracing messages explain the different situations.

I could add an assert.

Reply:
I would be happy if we can remove lengthy comment by using common ref-merge helper, but I don't have a strong opinion. The current implementation is also simple (except that we have to learn git2 API a bit more.)

Reply:
> I would be happy if we can remove lengthy comment by using common ref-merge helper, but I don't have a strong opinion.

+1 (both that it would be nice to avoid the long comment, and that I don't feel strongly)

Reply:
I'm not sure what I think at this point. As I alluded to, I do use the common ref-merge helper in tests, but I'm not sure how much of that (or the long comment) is actually helpful and how much is more of a historical artifact of me figuring out how things should work.

Reply:
Just for reference, this is my attempt to switch to be using merge_ref_targets(). Tests passed, but there might be some bugs.
```rust
                let oid_to_target = |oid: git2::Oid| {
                    let id = (!oid.is_zero()).then(|| CommitId::from_bytes(oid.as_bytes()));
                    RefTarget::resolved(id)
                };
                let remote_target = oid_to_target(update.src());
                if remote_target.added_ids().all(|id| repo.index().has_id(id)) {
                    // remote known, simulate merge
                    let base_target = RefTarget::resolved(expected_location.cloned());
                    let local_target = oid_to_target(update.dst());
                    let merged_target = refs::merge_ref_targets(
                        repo.index(),
                        &remote_target,
                        &base_target,
                        &local_target,
                    );
                    if merged_target != local_target {
                        failed_push_negotiations_ref.push(dst_refname.to_string());
                    }
                } else {
                    // remote unknown, must fetch
                    failed_push_negotiations_ref.push(dst_refname.to_string());
                }
```

Reply:
Thanks, this is close to what I was imagining and what I do in tests.

I still don't know what I think, but I don't consider this a blocker, my feelings about this are also not that strong. If I don't have any more concrete thoughts before we're done with all the other issues, I'll just use your piece of code and we can call it a day.

Reply:
I'm now thinking that if I change this to Yuya's way (which I'm still leaning towards), I'll add a comment and/or rework tests to make it clear what kind of fast-forwards are allowed (the logic I currently have written out explicitly). I just need to try it and see how it goes.

Reply:
I have a version of this, see the diff from the previous version here: https://github.com/ilyagr/jj/commit/fwl-mergerefs. It's longer than Yuya's since I again tried to explain both ways to understand the logic, this time more in code than in comments. The tests got a lot shorter, though.

I'll probably switch to it, or perhaps try to simplify it some more, but let me know if you have any thoughts.